lib/ecto_ldap/adapter.ex

```diff
@@ -230,11 +177,81 @@ defmodule Ecto.Ldap.Adapter do
   def translate_value(%Ecto.Query.Tagged{value: value}), do: value
   def translate_value(other), do: other
 
-  def load(:id, value), do: {:ok, value}
+  def execute(_repo, query_metadata, prepared, params, preprocess, _options) do
+
+    search_response   = search(prepared, params)
+    fields            = ordered_fields(query_metadata.sources)
+    count             = count_fields(query_metadata.select.fields, query_metadata.sources)
+
+    {:ok, {:eldap_search_result, results, []}} = search_response
+
+    result_set = results
+    |> Enum.map(&create_rows(&1))
+    |> Enum.map(&prune_columns(&1, fields))
+    |> Enum.map(&generate_models(&1, preprocess, count))
+
+    {count, result_set}
+  end
+
+  def ordered_fields(sources) do
+    {_, model} = elem(sources, 0)
+    model.__schema__(:fields)
+  end
+
+  def count_fields(fields, sources) do
+    Enum.map fields, fn
+      {:&, _, [idx]} = field ->
+        {_source, model} = elem(sources, idx)
+        {field, length(model.__schema__(:fields))}
+      field ->
+        {field, 0}
+    end
+  end
+
+  def create_rows({:eldap_entry, dn, attributes}) when is_list(attributes) do
+    List.flatten(
+      [dn: dn], 
+      Enum.map(attributes, fn {key, value} ->
+        {key |> to_string |> String.to_atom, value}
+      end))
+  end
+
+  def prune_columns(attributes, fields) do
+    fields
+    |> Enum.map(fn column -> Keyword.get(attributes, column) end)
+  end
+
+  def generate_models(row, preprocess, fields) do
+    Enum.map_reduce(fields, row, fn
+      {field, 0}, [h|t] ->
+        {preprocess.(field, h, nil), t}
+      {field, count}, acc ->
+        case split_and_not_nil(acc, count, true, []) do
+          {nil, rest} -> {nil, rest}
+          {val, rest} -> {preprocess.(field, val, nil), rest}
+        end
+    end) |> elem(0)
+  end
+
+  def split_and_not_nil(rest, 0, true, _acc), do: {nil, rest}
+  def split_and_not_nil(rest, 0, false, acc), do: {:lists.reverse(acc), rest}
+
+  def split_and_not_nil([nil|t], count, all_nil?, acc) do
+    split_and_not_nil(t, count - 1, all_nil?, [nil|acc])
+  end
+
+  def split_and_not_nil([h|t], count, _all_nil?, acc) do
+    split_and_not_nil(t, count - 1, false, [h|acc])
+  end
+
+  def load(:id, value),     do: {:ok, value}
   def load(:string, value), do: {:ok, convert_from_erlang(value)}
 
-  def convert_from_erlang(list=[head|_]) when is_list(head), do: Enum.map(list, &convert_from_erlang/1)
-  def convert_from_erlang(string) when is_list(string), do: :binary.list_to_bin(string)
-  def convert_from_erlang(other), do: other
+  def dump(:string, value), do: {:ok, convert_to_erlang(value)}
+  def dump(_, value),       do: {:ok, convert_to_erlang(value)}
+
+  def convert_from_erlang(list=[head|_])  when is_list(head),   do: Enum.map(list, &convert_from_erlang/1)
+  def convert_from_erlang(string)         when is_list(string), do: :binary.list_to_bin(string)
+  def convert_from_erlang(other),                               do: other
```

Comment:
Harder to read because I don't need to compare implementations, but do need to compare function heads.